### DISCUSSION TITLE
Touch Support Incomplete in EDK2?

### DISCUSSION THREAD

**Discussion by CalculonPrime** [tianocore/edk2#2755] [Q&A]
I'm working with software (VeraCrypt) that uses EDK2 to support touch interaction at boot.  It doesn't seem to work with the ASUS Vivobook 14 Flip, however, so I rebuilt the app with some debugging code.
 
There is a TouchDevice detected and a handle is obtained, but when DevicePathFromHandle() is called, it returns 3 (ERROR_UNSUPPORTED).  Is DevicePathFromHandle() not returning success directly related to why touch isn't working, or can touch work without this function returning a valid path?
 
I'm thinking that perhaps EDK2's touch support may be incompatible with the Vivobook 14.
 
Note that VeraCrypt is built off the UDK2015 branch.  Does that branch not have reliable support for touch?

**Comment by spbrogan** [comment]
This really isn't a support forum for any commercially available product.  Also note that proprietary firmware (that which is released for a given product) doesn't usually use edk2 as seen exactly in this repository.  Vendors generally make changes for their products and have significant code coming from other sources (IBV, ODM, OEM, silicon vendors and partners).  Those same vendors generally do not publicly share the source for a given product.  Also note you are asking about a 7 year old branch.  This project runs on a mainline development branch and has very little support for "backport" or support for older tags and branches. 

Finally, touch support is generally custom/embedded per product.   Many solutions require custom power on/off sequences and custom drivers to translate into UEFI input protocols like absolute pointer.   In most operating systems they leverage custom drivers or at least custom ACPI code to support "standardized" HID.  

If you wanted to try to replicate your scenario using one of the edk2 or edk2-platform open-source repositories running tip of tree you might get better feedback.  

Good luck.

**Reply by CalculonPrime** [reply]
As mentioned, the reason I'm building against UDK2015 is because the problem is with VeraCrypt-DCS, and EDK2-based boot-time software app.  I have no control over what that project does, and no ability to push them to upgrade to a more recent EDK2.  I created issues on their Github page reporting the problem, but so far they've gone unanswered.

If I wanted to see whether UDK2015 is causing the issue (i.e., using an old release), I'd need to separate my test effort from VeraCrypt-DCS.  Is there a "hello world" test app included in EDK2 that can be built that uses touch events?

**Reply by spbrogan** [reply]
I am not aware of one.  In the platforms I have seen, touch is turned in absolute pointer (https://github.com/tianocore/edk2/blob/master/MdePkg/Include/Protocol/AbsolutePointer.h) .  So any app that uses absolute pointer could be considered a way to test your touch support.  But again, I view touch support is an internal "platform supplied thing" and thus not the responsibility of the application layer to supply.  Seems that VeraCrypt tried to enable touch support so I would suggest working with that project.

**Reply by CalculonPrime** [reply]
The VeraCrypt-DCS owner is non-responsive to the issues I have logged.  My top-level goal is to determine whether the bug is in the VeraCrypt-DCS code, or EDK2.  So, that's why I was looking for a "hello world" program that demonstrates touch screen events.  I'm actually surprised that there isn't a test application shipping with EDK2 that features touch screen events.

Is there any project known to use EDK2 which supports touch screen events, and supports building on Windows with Visual Studio?  That's how VeraCrypt DCS is built.  I found an EDK2-based project called rEFind that supports touch events, but it doesn't support building on Windows.  If I can't build it in the same environment, then even if it works I won't know whether that means the problem is in VeraCrypt-DCS or Windows build environment.

**Reply by CalculonPrime** [reply]
BTW, where is the correct API documentation for using EDK2 in applications?  I found a "driver guide" that documents, handleProtocol(), for example, but it says not to use it, as it's not right for drivers.  Well, what about in an application, not a driver?  Where can I find documentation for using handleProtocol() and locateHandle() in an application?

**Reply by lersi** [reply]
in the readme of the following repo, I have gathered a lot of links that explains different parts of EDK2, including user app.

repo link: [edk2_project_template](https://github.com/lersi/edk2_project_template/)
